### PR TITLE
Devtooling 272- Exclude Attributes and multilevel dependency resolve

### PR DESCRIPTION
**Pull request by HemanthDogiparthi12**

### CODE REVIEW

**Comment by charliecon** [comment]
@HemanthDogiparthi12 Is this ready for review? I tried it out locally and it didn't work for me. Also the matrix tests caught some errors in `genesyscloud/tfexporter/genesyscloud_resource_exporter_test.go`

File: genesyscloud/tfexporter/export_common.go [hunk 1]
```diff
@@ -88,6 +88,30 @@ func FilterResourceByName(result resourceExporter.ResourceIDMetaMap, name string
 	return result
 }
 
+func FilterResourceById(result resourceExporter.ResourceIDMetaMap, name string, filter []string) resourceExporter.ResourceIDMetaMap {
```
Comment (carnellj-genesys) [inline genesyscloud/tfexporter/export_common.go:91]: This feels like it should have a a couple of unit tests to provide this out.

File: genesyscloud/tfexporter/genesyscloud_resource_exporter.go [hunk 2]
```diff
@@ -499,19 +517,95 @@ func (g *GenesysCloudResourceExporter) exportDependentResources(filterList []str
...
+	g.buildResourceConfigMap()
+	g.exportAndResolveDependencyAttributes()
+
+	g.resources = append(existingResources, append(uniqueResources, g.resources...)...)
+	g.exporters = mergeExporters(existingExporters, *mergeExporters(depExporters, *g.exporters))
+
+	return nil
+}
+
+func (g *GenesysCloudResourceExporter) exportAndResolveDependencyAttributes() (diagErr diag.Diagnostics) {
```
Comment (carnellj-genesys) [inline genesyscloud/tfexporter/genesyscloud_resource_exporter.go:545]: Unit tests ? Either was fricking nice :)

File: genesyscloud/tfexporter/genesyscloud_resource_exporter.go [hunk 3]
```diff
@@ -499,19 +517,95 @@ func (g *GenesysCloudResourceExporter) exportDependentResources(filterList []str
...
+
+	if len(filterListById) > 0 {
+		g.resourceFilter = FilterResourceById
+		g.chainDependencies(make([]resourceExporter.ResourceInfo, 0), exp)
+	}
+
+	return nil
+}
+
+// Recursive function to perform operations based on filterListById length
```
Comment (carnellj-genesys) [inline genesyscloud/tfexporter/genesyscloud_resource_exporter.go:570]: unit test?

File: genesyscloud/tfexporter/genesyscloud_resource_exporter.go [hunk 4]
```diff
@@ -1079,11 +1184,61 @@ func populateConfigExcluded(exporters map[string]*resourceExporter.ResourceExpor
...
 		}
+
 		excludedAttr := excluded[resourceIdx+1:]
 		exporter.AddExcludedAttribute(excludedAttr)
 		log.Printf("Excluding attribute %s on %s resources.", excludedAttr, resourceName)
 	}
 	return nil
 }
+
+func (g *GenesysCloudResourceExporter) resolveReference(refSettings *resourceExporter.RefAttrSettings, refID string, exporters map[string]*resourceExporter.ResourceExporter, exportingState bool) string {
```
Comment (carnellj-genesys) [inline genesyscloud/tfexporter/genesyscloud_resource_exporter.go:1204]: Unit tests?

**Review by carnellj-genesys — COMMENTED** [review]
Overall very nice work.  However, with some of these new functions, I would spend the extra time doing some unit tests.  We are getting testing as a by product of our export acceptance tests.  I would like to see test cases that can provide this code works not because I don't believe that you have tested but rather, we are going to be a doing a lot more work and refactoring of this code and I think the regression tests are important.

File: genesyscloud/tfexporter/genesyscloud_resource_exporter.go [hunk 5]
```diff
@@ -1003,6 +1105,7 @@ func removeZeroValues(key string, val interface{}, configMap gcloud.JsonMap) {
 // Identify the parent config map and if the resources have further dependent resources add a new attribute depends_on
 func (g *GenesysCloudResourceExporter) addDependsOnValues(key string, configMap gcloud.JsonMap) {
 	list, exists := g.dependsList[key]
+	//configMap["key"] = key
```
Comment (charliecon) [inline genesyscloud/tfexporter/genesyscloud_resource_exporter.go:1108]: leftover comment

**Review by charliecon — APPROVED** [review]

**Review by carnellj-genesys — APPROVED** [review]

**Pull request merged**